idgen/idgen.go

```diff
@@ -0,0 +1,147 @@
+package idgen
+
+import (
+	"fmt"
+	"math/bits"
+	"sync"
+
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	kbucket "github.com/libp2p/go-libp2p-kbucket"
+)
+
+// HydraIdentityGenerator is a shared balanced ID generator.
+var HydraIdentityGenerator = NewBalancedIdentityGenerator()
+
+// BalancedIdentityGenerator is a facility for generating IPFS identities (i.e. IPFS private keys),
+// whose corresponding DHT keys are highly balanced, compared to just generating random keys.
+// Balancing is accomplished using "the power of two choices" paradigm:
+// https://www.eecs.harvard.edu/~michaelm/postscripts/mythesis.pdf
+//
+// New identities are generated by calling AddBalanced. BalancedIdentityGenerator remembers
+// generated identities, in order to ensure balance for future identities.
+// Generated identities can be removed using Remove.
+//
+// BalancedIdentityGenerator maintains the invariant that all identities, presently in its memory,
+// form an almost-perfectly balanced set.
+type BalancedIdentityGenerator struct {
+	sync.Mutex
+	xorTrie *XorTrie
+	count   int
+}
+
+// NewBalancedIdentityGenerator creates a new balanced identity generator.
+func NewBalancedIdentityGenerator() *BalancedIdentityGenerator {
+	return &BalancedIdentityGenerator{
+		xorTrie: NewXorTrie(),
+	}
+}
+
+// AddUnbalanced is used for testing purposes. It generates a purely random identity,
+// which is not balanced with respect to the existing identities in the generator.
+// The generated identity is stored in the generator's memory.
+func (bg *BalancedIdentityGenerator) AddUnbalanced() (crypto.PrivKey, error) {
+	bg.Lock()
+	defer bg.Unlock()
+	p0, t0, _, err0 := bg.genUniqueID()
+	if err0 != nil {
+		return nil, fmt.Errorf("generating unbalanced ID candidate, %w", err0)
+	}
+	bg.xorTrie.Insert(t0)
+	bg.count++
+	return p0, nil
+}
+
+// AddBalanced generates a random identity, which
+// is balanced with respect to the existing identities in the generator.
+// The generated identity is stored in the generator's memory.
+func (bg *BalancedIdentityGenerator) AddBalanced() (crypto.PrivKey, error) {
+	bg.Lock()
+	defer bg.Unlock()
+	p0, t0, d0, err0 := bg.genUniqueID()
+	if err0 != nil {
+		return nil, fmt.Errorf("generating first balanced ID candidate, %w", err0)
+	}
+	p1, t1, d1, err1 := bg.genUniqueID()
+	if err1 != nil {
+		return nil, fmt.Errorf("generating second balanced ID candidate, %w", err1)
+	}
+	if d0 < d1 {
+		bg.xorTrie.Insert(t0)
+		bg.count++
+		return p0, nil
+	} else {
+		bg.xorTrie.Insert(t1)
+		bg.count++
+		return p1, nil
+	}
+}
```

Comment:
This is such an elegant solution. I'm trying to conceive it all in my head and I got some questions:

- What happens when adding a PeerId doesn't not increase the total depth?
- Why aren't you adding the PeerIds that are seen in the network? If you were to add them to the trie, would that impact your decision? 
  - My understanding is that this func only optimizes for speeding up the process of creating an even distribution of PeerIds, however, it won't be able to detect hotspots in the address space and inject new Peers there. Am I getting this correctly? 
- What would it take to be able to generate Ids closer to hot spots (would it just be picking the Ids that lead to less depth? How would we differentiate the sybils from the other nodes here)? This would enable us to have reactive hydra-nodes

Reply:
> What happens when adding a PeerId doesn't not increase the total depth?

Short answer: It doesn't matter. In fact, this is desired: If an insertion does not increase the total depth, it means it fell in a "hole" where the tree was shallower. Thus making the tree more balanced.

Note that the algorithm does not even look at the total depth. The algorithm uses a probabilistic phenomenon: If the current tree is unbalanced, then with good probability one of the two candidate IDs will fall in a shallower part of the tree, and thus will fill a hole. It is a non-trivial theorem (hence the link to a paper) that if you follow this procedure, the tree will always be approximately balanced: The average depth will be log N, while the maximum depth will be
log N + log log N. In comparison, if you just pick random IDs, the tree will have average depth
of a bit more than log N, and a maximum depth of 6 log N.

Why does depth reflect density:
The depth where an ID falls when inserted represents how close it is to existing nodes. If a node is inserted at depth D, then its closest node is at a distance 2^(-D). Having all nodes at similar depth means they are equidistant in the XOR metric. 

> Why aren't you adding the PeerIds that are seen in the network?  ...

You are getting this correctly. If you only insert the Hydra nodes in the trie, then you guarantee that the Hydra nodes cover the ID space evenly. (The other nodes may be uneven.)

We could also decide to add the other "seen" nodes to the trie. This will have the effect of evening out the distribution of all nodes in the system, with Hydra nodes being denser in places where other nodes are not as dense.

I would have a design discussion before taking this step, for two reasons:
* One can argue that KadDHT should itself be changed so that normal DHT nodes follow a balanced insertion policy in the first place using a similar technique (this can be done).
*  A "hot spot" (in my vocabulary) is not synonymous with a place in the ID space where there is a lower density of DHT nodes. A "hot spot" is a place that has content of high demand. That's why I  opted to balance Hydra nodes with respect to the ID space (which represents all content), and not with the specific distribution of other nodes (which represents nothing more than the randomness in ID choices.)

> What would it take to be able to generate Ids closer to hot spots ...

What is your definition of a "hot spot"?